drivers/display/display_st7789h2.c

```diff
@@ -0,0 +1,647 @@
+		if (dev_cur != dev_pin || pin != (lastpin+1)) {
+			data->port_groups_count++;
+			dev_cur = config->data_gpios[c].port;
+			data->port_groups[data->port_groups_count].port = dev_cur;
+			data->port_groups[data->port_groups_count].len = 0;
+			data->port_groups[data->port_groups_count].offset = pin;
+		}
+
+		data->port_groups[data->port_groups_count].len++;
+
+		lastpin = pin;
+	}
+	data->port_groups_count++;
+
+	LOG_INF("data_gpios optimized");
+#endif
+
+	st7789h2_power_up(config);
+
+	st7789h2_backlight_on(config);
+
+	st7789h2_reset_display(dev);
+
+	st7789h2_blanking_on(dev);
+
+	st7789h2_lcd_init(dev);
+
+	st7789h2_exit_sleep(dev);
+
+	return 0;
+}
+
+#ifdef CONFIG_PM_DEVICE
+static int st7789h2_pm_control(const struct device *dev,
+			      enum pm_device_action action)
+{
+	int ret = 0;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		st7789h2_exit_sleep(dev);
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		st7789h2_transmit(dev, ST7789V_CMD_SLEEP_IN, NULL, 0);
+		break;
+	default:
+		ret = -ENOTSUP;
+		break;
+	}
+
+	return ret;
+}
+#endif /* CONFIG_PM_DEVICE */
+
+static const struct display_driver_api st7789h2_api = {
+	.blanking_on = st7789h2_blanking_on,
+	.blanking_off = st7789h2_blanking_off,
+	.write = st7789h2_write,
+	.read = st7789h2_read,
+	.get_framebuffer = st7789h2_get_framebuffer,
+	.set_brightness = st7789h2_set_brightness,
+	.set_contrast = st7789h2_set_contrast,
+	.get_capabilities = st7789h2_get_capabilities,
+	.set_pixel_format = st7789h2_set_pixel_format,
+	.set_orientation = st7789h2_set_orientation,
+};
+
+#ifdef CONFIG_ST7789H2_GPIO_PORT_WRITING
+#define ST7789H2_DATA_INIT(inst) \
+	struct st7789h2_gpio_port_group                                                            \
+		st7789h2_data_port_groups_ ## inst[ST7789H2_DATA_NGPIOS(inst)];                    \
+	static struct st7789h2_data st7789h2_data_ ## inst = {                                     \
+		.width = DT_INST_PROP(inst, width),                                                \
+		.height = DT_INST_PROP(inst, height),                                              \
+		.x_offset = DT_INST_PROP(inst, x_offset),                                          \
+		.y_offset = DT_INST_PROP(inst, y_offset),                                          \
+		.port_groups = st7789h2_data_port_groups_ ## inst,                                 \
+		.port_groups_count = ST7789H2_DATA_NGPIOS(inst),                                   \
+	};
+#else
+#define ST7789H2_DATA_INIT(inst) \
+	static struct st7789h2_data st7789h2_data_ ## inst = {                                     \
+		.width = DT_INST_PROP(inst, width),                                                \
+		.height = DT_INST_PROP(inst, height),                                              \
+		.x_offset = DT_INST_PROP(inst, x_offset),                                          \
+		.y_offset = DT_INST_PROP(inst, y_offset),                                          \
+	};
+#endif
```

Comment:
```suggestion
#ifdef CONFIG_ST7789H2_GPIO_PORT_WRITING
#define PORT_GROUPS_DEFINE(inst)                                                                   \
	static struct st7789h2_gpio_port_group                                                     \
		st7789h2_data_port_groups_ ## inst[ST7789H2_DATA_NGPIOS(inst)]                     \
#define PORT_GROUPS_INIT(inst)                                                                     \
		.port_groups = st7789h2_data_port_groups_ ## inst,                                 \
		.port_groups_count = ST7789H2_DATA_NGPIOS(inst),
#else
#define PORT_GROUPS_DEFINE(inst)
#define PORT_GROUPS_INIT(inst)
#endif

#define ST7789H2_DATA_INIT(inst)                                                                   \
	PORT_GROUPS_DEFINE(inst);                                                                  \
	static struct st7789h2_data st7789h2_data_ ## inst = {                                     \
		.width = DT_INST_PROP(inst, width),                                                \
		.height = DT_INST_PROP(inst, height),                                              \
		.x_offset = DT_INST_PROP(inst, x_offset),                                          \
		.y_offset = DT_INST_PROP(inst, y_offset),                                          \
		PORT_GROUPS_INIT(inst),                                                            \
	};
```

(and possibly just drop ST7789H2_DATA_INIT/ST7789H2_CONFIG_INIT, and put its content directly in the definition macro)

Reply:
Nice alternative thanks :)

Reply:
Unfortunately, I tried that but it broke the other macros so I'll stick to my solution for now